CL2-6900 Fix URL generation from inappropriate content notification

jinjagit opened the pull request:
## Checklist

- [x] Added entry to changelog
<details>
<summary>More info</summary>
Add a concise line to the 'Next release' section of the changelog (docs/README.md) so people other than developers can understand what has changed where. E.g. 'Added an error message to the project name field of the project edit form (Admin > Projects > Edit)'.
</details>

- [x] WCAG 2.1 AA proof
<details>
<summary>More info</summary>
For front-end devs only. Is your work conforming with the WCAG 2.1 AA rules? If you need more info, read the [a11y page](https://www.notion.so/citizenlab/a11y-7568f83d42ab4895ac133b89d358997b) on our Notion.
</details>

- [x] Tests
<details>
<summary>More info</summary>

### Unit tests

Did you add relevant unit tests?

### E2E tests

Sometimes it can be more efficient to update E2E tests after CI has run them. If you know which ones to update, go ahead! E2E template cl2-back:

```bash
docker compose run --rm web bin/rails cl2_back:create_tenant[localhost,e2etests_template]
```

</details>

- [x] Prepared branch for code review
<details>
<summary>More info</summary>
Reviewed code to reduce unnecessary back and forth (removal of console.log, comments, ...)? Added comments to clarify code, emphasize what to pay attention to, etc.?
</details>

## How urgent is a code review?

With Xmas fast approaching, by the end of the week would be nice :)

jinjagit committed:
Add test - failing as expected

Currently failing as we are not yet returninga relative path to
the respective Idea, but rather an absolute URL = the bug.

jinjagit committed:
Rough fix - BE now passes flaggable_path, not _url, to FE

Test commented out to see what else fails on CI.
Probably now includes some redundant code, to be removed.

jinjagit committed:
Set test to red

Weirdly didn't appear to fail CI when first pushed, so let's
check it does now.

jinjagit committed:
Remove redundant code + set test to green

Removed flaggable_url from InappropriateContentFlaggedSerializer.
Removed flaggable_url type definition from notifications.d.ts.
Used flaggable_path in test = green.

jinjagit commented on back/engines/free/frontend/app/services/frontend/url_service.rb:
```diff
@@ -4,7 +4,7 @@ module Frontend
   # makes about the frontend URLs into a single location.
   class UrlService
-    def model_to_url(model_instance, options = {})
+    def model_to_path(model_instance)
```
I chose to refactor `def model_to_url` into 2 defs: `def model_to_path` which is now used by `def model_to_url`.

This should mean `def model_to_url` continues to function as previously, whilst we can now use `def model_to_path` to fix this bug (and for any other case where we need to give the FE a path, not a full URL).

I could have created a new `PathService` but chose not to, primarily because of the comment in this file:
```
# The main purpose of this service is to decouple all assumptions the backend
# makes about the frontend URLs into a single location.
```

It could also be argued that this refactoring avoids the code duplication involved in creating a new, very similar service.

I am aware that a path is not, strictly speaking, a URL, but on balance still feel my approach makes sense.

Do you agree/disagree?

kogre replied:
I think it's a great solution.

jinjagit committed:
Update changelog

jinjagit committed:
Fix unrelated typo that affected emails re. flagged content

The mis-spelling of flaggable_author_name probably meant that
either the value was never displayed on emails regarding flagged
content, or possibly that the email would not be created (failed).

jinjagit commented on front/app/modules/commercial/flag_inappropriate_content/citizen/components/NLPFlagNotification.tsx:
```diff
@@ -19,7 +19,7 @@ const NLPFlagNotification = memo<Props>((props) => {
   return (
     <NotificationWrapper
-      linkTo={notification.attributes.flaggable_url}
+      linkTo={`/${notification.attributes.flaggable_path}`}
```
I chose to add the leading `/` here, rather than in the BE `def model_to_path`, since conventionally relative paths do not start with a leading `/`. However, I have no strong feelings for or against either approach.

kogre replied:
Thanks for pointing that out, I might have missed it.

I think you should make it start with a slash, because relative paths (the way we are using them here) are _not_ really relative paths, but apparently called `absolute-path references`. If found that nice term in [this SO thread](https://webmasters.stackexchange.com/questions/56840/what-is-the-purpose-of-leading-slash-in-html-urls).

jinjagit commented on back/engines/commercial/flag_inappropriate_content/app/models/flag_inappropriate_content/email_campaigns/campaigns/inappropriate_content_flagged.rb:
```diff
@@ -72,7 +72,7 @@ def generate_commands recipient:, activity:, time: nil
         payload = {
           flaggable_type: data[:flaggable_type],
-          flaggalbe_author_name: UserDisplayNameService.new(AppConfiguration.instance, recipient).display_name!(data[:flaggable_author]),
+          flaggable_author_name: UserDisplayNameService.new(AppConfiguration.instance, recipient).display_name!(data[:flaggable_author]),
```
Not directly related to this PR, but seemed an obvious typo in need of fixing.

jinjagit committed:
Obey rubocop

kogre commented on back/engines/commercial/flag_inappropriate_content/spec/serializers/flag_inappropriate_content/web_api/v1/notifications/inappropriate_content_flagged_serializer_spec.rb:
```diff
@@ -0,0 +1,13 @@
+    hash = described_class.new(flagged).serializable_hash
+    slug = Idea.find(flagged.inappropriate_content_flag.flaggable.post_id).slug
+
+    expect(hash[:data][:attributes][:flaggable_path]).to eq("ideas/#{slug}")
```
This is not wrong, happy you covered it with a spec, but just a small piece of advice:

It's often better to not use variables in your expectation checks when you can avoid them. In this case, by checking it equals `ideas/#{slug}` your test become a bit less robust. Suppose for example the inappropriate_content_flagged model changes in some way, and we rename `slug` to `plug`. This test would still succeed, as it would render as `ideas/nil`.

The alternative is that you actively set the slug to a specific string when you create the data, and then check that exact string in your assertion, that will never succeed by coincidence.

jinjagit replied:
Good catch. I completely forgot about the `nil` thing. Thanks! 👍

kogre reviewed (changes requested):
Nice solution!

Would just add the leading slash, see comments, otherwise all good.

jinjagit committed:
Use string, not variable, in test assertion

jinjagit committed:
Return path with leading forward slash from def model_to_path

This removes need to add the forward slash in the FE, when making
use of the return value(s) of this method.
